Add Dropdown component

jgiardino opened the pull request:
This adds the dropdown component as described in issue #23.

closes #23

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit ecedbe0c3dff313cf85b9727400fa329595594d8

https://deploy-preview-206--pf-next.netlify.com

jgiardino commented:
@kybaker This is ready for a visual design review. Could you take a look and let me know if anything doesn't look how you expect it to?

@srambach I'm curious what your thoughts are for the [dropdown arrow](https://github.com/patternfly/patternfly-next/pull/206/files#diff-e03e44eb3caae36bfc8247225a6ab737R73), and if we should use css to create elements like that, or if they should be based on icons. 

Thanks :smile:

kybaker commented:
@jgiardino It looks good. My only concern is that the menu is using the largest shadow variables. Hard to tell from the inspector.

jgiardino commented:
Thanks, @kybaker!

For the box shadow, I used [$pf-global--BoxShadow](https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/utilities/variables.scss#L68). I assumed that the absence of a size in the name meant it was the default one to use unless otherwise specified, but that could totally be a wrong assumption (or maybe the variable name is incorrect and it should really be $pf-global--BoxShadow--md?).

Here is a screen capture using the next and only size down, $pf-global--BoxShadow--sm:

![image](https://user-images.githubusercontent.com/21063328/39828870-8b386dd4-538a-11e8-836d-1a6462d8607c.png)

Let me know if that's the variable I should use (if not, then we might need a separate issue to update box shadow variables).

kybaker commented:
@jgiardino The previous shadow was better. Thanks!

srambach commented:
@jgiardino, on the carets -  I guess one question would be, do we ever want to change, or allow someone to change, the caret to something else? With the CSS solution, it's pretty tied to being what it is. We are parameterizing colors, spacing, etc., but this would preclude doing that for the icon. I'm not saying it's bad or good, just that it would be a limitation. Other than that, I can't really think of anything.

mcarrano commented:
@jgiardino @kybaker In thinking about this a bit more, we probably want to account for disabled menu items as this is a pretty common use case when dropdowns are used for invoking actions.

jgiardino commented:
@mcarrano - 👍 sure, I'll include this in my next commit

@andresgalante - do you have any suggestions regarding @srambach's comment about the arrow? We don't have icons yet, which is the main reason I'm using the css solution. But if we use an icon, I'm also not sure how to parameterize that.

jgiardino commented:
@kybaker Do you have visual design details for how disabled menu items should display?

kybaker commented:
@jgiardino The text should just use a lighter hex value. We have not defined a disabled hex value yet. When I have it I will send it over - we can always add that visualization later.

jgiardino commented:
The latest update includes the following:

- examples for dropup, right-aligned, kebab (i.e. ghost + sans arrow)
- updated html for better accessibility 
    - @seanforyou23, @markcaron let me know if you have any feedback on this one
- included documentation
- styles for the examples sections to: 
    - set the height based on the height of the dropdown to avoid overlapping the html sample; this includes a separate class for the dropup example
    - set the alignment to right for the right aligned menu examples
    - fyi @dgutride ^^^

There are still some open issues that are likely to change:

- the text [kebab] will need to be replaced with an icon when one is available
- the arrow will likely get replaced with an icon? (I'd like @kybaker's input on this one)
- the way we handle the thicker border on hover will likely change (I think @mattnolting is using a different technique in forms that we'll use here)

There are a couple of things I was guessing at that I'd like input on (@mcarrano @kybaker):

- For the Dropup button menu, which direction should the arrow point when the menu is collapsed?
- Styles used for disabled menu items

mcarrano commented:
@jgiardino regarding the "DropUp" what do you think the intended usage is for this? I guess I would expect it to appear as you have it in the preview. As a designer, I wouldn't treat this any differently than a dropdown, but would expect the code to be smart enough to open the menu above if there is not enough space to drop the menu below the control.

jgiardino commented:
@mcarrano 
What you describe matches how I was thinking of it, too.

The use case I was imagining was on a page like card view or list view, where each item has a dropdown where the menu drops down, but the items at the end of the page would display the menu above the toggle to account for not having space below the toggle. I assume that would happen dynamically, hence the down arrow in the collapsed state so that they all have a consistent appearance in their initial state. I don't think we'd statically set `.pf-is-dropup` except in cases like a pagination row that we know would always be at the bottom.

andresgalante commented on src/patternfly/components/Dropdown/examples/dropdownKebab.hbs:
```diff
@@ -0,0 +1 @@
+{{#> dropdown id="dropdown-example-kebab" expanded="true" styleModifier="pf-is-ghost pf-sans-arrow"}}[Kebab]{{/dropdown}}
```
what happened with the 💩 ?

jgiardino replied:
I was afraid I might step in it. 
![bitmoji](https://render.bitstrips.com/v2/cpanel/ca46077b-19a4-41d8-95d2-0bc33146e21c-fe7ba5ad-9c01-4dac-b094-998c3f204a3b-v1.png?transparent=1&palette=1&width=123)

christianvogt commented:
The first and last elements in the list have odd padding that make the text appear misaligned. I suggest keeping the menu items the same size but adding padding to the menu itself so that these menu items don't look different than the rest.
![image](https://user-images.githubusercontent.com/14068621/40194260-c9224b8a-59d7-11e8-81fd-c30a64028fdc.png)

christianvogt commented on src/patternfly/components/Dropdown/dropdown.hbs:
```diff
@@ -0,0 +1,13 @@
+    <a class="pf-c-dropdown__menu-item pf-is-disabled" role="menuitem" aria-disabled="true" tabindex="-1" href="#">Disabled Link</a>
+    <button class="pf-c-dropdown__menu-item pf-is-disabled" role="menuitem" disabled="disabled">Disabled Action</button>
+    <div class="pf-c-dropdown__separator" role="separator"></div>
+    <a class="pf-c-dropdown__menu-item" href="#">Separated Link</a>
```
Missing role `menuitem`.

christianvogt commented on src/patternfly/components/Dropdown/dropdown.hbs:
```diff
@@ -0,0 +1,13 @@
+<div class="pf-c-dropdown{{#if expanded}} pf-is-expanded{{/if}} {{styleModifier}}">
+  <button id="{{id}}" class="pf-c-dropdown__toggle" aria-haspopup="true" aria-expanded="{{#if expanded}}true{{else}}false{{/if}}">
```
Should we also be using `aria-controls` and reference the menu's id? I know in the past `aria-controls` had questionable support from screen readers.

jgiardino replied:
When I tested `aria-controls` it seems best suited for elements that are not adjacent to the controlling element in the dom. In JAWS, it provides a way to shift focus to that element, and to me seems comparable to an anchor link to another element on the page. I think a better use case for it would be things like toggles or filters on a toolbar that change the view below it.

Something else worth noting, if `aria-controls` is associated with a hidden element, then JAWS will not announce anything related to the `aria-controls` attribute. 

When we reviewed menus with screen reader users, one piece of feedback we received is that focus should automatically be placed on the first menu item. But if it isn't, they had no trouble finding the menu items associated with the menu button. If the dropdown toggle says that it's expanded, they assume the dropdown menu items are the next items that would be announced.

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+    background-color: $pf-c-dropdown__toggle--BackgroundColor;
+    border: $pf-c-dropdown__toggle--BorderWidth solid $pf-c-dropdown__toggle--BorderColor;
+
+    &:hover {
```
Button also uses `.pf-is-hover` and `.pf-is-active`. What's the convention to follow?

jgiardino replied:
@andresgalante or @srambach Do you have thoughts on this one? My understanding of these classes in the Button component was so that we can display all the variations of the button without actually having to hover or click on them.

Also, instead of `.pf-is-active` I use `.pf-is-expanded` which for the Dropdown are one in the same, but "expanded" seems more explicit about the state.

christianvogt replied:
@jgiardino is the expectation that on mouse down the menu is expanded? If not, then `active` is a different state. The bootstrap dropdown menu doesn't open on mouse down and has an `active` state just like a normal button. Actually at this point I'm a bit confused as to why we wouldn't simply utilize a the PF button as the toggle button for dropdown thereby allowing for all the different variants.

jgiardino replied:
I'm not sure I follow what you mean—when I look at the [bootstrap dropdown](https://getbootstrap.com/docs/4.1/components/dropdowns/#single-button) example, the menu is visible when I click on the toggle. And the styles applied when the toggle button is `:active` are the same styles applied when the menu is visible. I agree, they're two different things. I think what I meant was that the styles we would apply to the toggle in the active state should also get applied when the menu is expanded. But if that is true, then should I also add a `.pf-c-dropdown__toggle:active` selector to the styles currently defined for the `.pf-is-expanded .pf-c-dropdown__toggle` selector (i.e two comma-separated selectors)?

christianvogt replied:
The bootstrap dropdown button can be `:active` without being expanded. It occurs on mouse down. You can mouse down on the button (active state), then move the mouse away before you mouse up which won't opened the menu. True that once the dropdown is opened, the button has the same active treatment.

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+      border-color: transparent;
+    }
+
+    .pf-sans-arrow & {
```
Is `sans` another modifier type? Or should `is` be always used?

christianvogt replied:
I believe @andresgalante agreed that PF should stick to a single consistent naming scheme for all modifiers. Having multiple variants for modifiers makes it difficult for creating utilities that can auto prefix modifier states. It would need too much smarts to know when to use `is` vs `has` vs `sans`.
`pf-is-noarrow` seems ok to me.

jgiardino replied:
That's a great point. I like `pf-is-noarrow`. Thanks for the suggestion!

andresgalante replied:
I don't like `sans` either. 

@christianvogt I want to build modifier with both `-is` or `-has`.

jgiardino commented:
@christianvogt That's a great question for @kybaker. I could have totally implemented that different from what the visual design team was thinking. 

@kybaker Can you provide some guidance on @christianvogt's comment above?

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+      border-width: $pf-c-dropdown__toggle--hover--BorderWidth;
+    }
+
+    .pf-is-expanded & {
```
I see the use of `.pf-is-modifier &` as bad. If PF starts to adopt this type of pattern you could end up with CSS from a parent inadvertently affecting child elements. I think PF should only ever use a modifier as a qualifier to a block or element or the rule must explicitly state the parent child relationship using `>`. eg. `.pf-is-expanded > &`

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+
+  &__menu {
+    position: absolute;
+    z-index: 1000;
```
Does PF not prescribe a set of elevation variables? If not, I think it's worth looking into.

jgiardino replied:
@andresgalante did you have something in mind that you'd like me to add?

Should I just do: `$pf-global--ZIndex:    1000;` for now, and we update them later?

andresgalante replied:
@jgiardino if you are doing that please create an issue for us to add zindex global variables later. If not just add them :)

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+      padding-bottom: $pf-c-dropdown__menu-item--last--PaddingBottom;
+    }
+
+    &:hover,
```
Button has different styles for hover, active, disabled and focus. PF should be consistent.

christianvogt commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+    }
+  }
+
+  &__menu {
```
Should menu be part of dropdown or a separate component altogether? A dropdown could hold many different types of information and a menu may be reusable elsewhere as well.

jgiardino replied:
I was wondering the same thing, about both breaking this out into separate components and allowing other things to display in what drops down besides a menu. But I would prefer to separate them when that need comes up, and when the designers have provided specific designs for those variations.

jgiardino commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+$pf-c-dropdown__toggle--PaddingRight: $pf-global--spacer--md !default;
+$pf-c-dropdown__toggle--PaddingBottom: $pf-global--spacer--xs !default;
+$pf-c-dropdown__toggle--PaddingLeft: $pf-global--spacer--sm !default;
+$pf-c-dropdown__toggle--FontSize: $pf-global--FontSize--md !default;
```
Thanks for catching that!

andresgalante commented on src/patternfly/components/Dropdown/styles.scss:
```diff
@@ -0,0 +1,158 @@
+
+    &:hover {
+      padding-top: calc(#{$pf-c-dropdown__toggle--PaddingTop} - (#{$pf-c-dropdown__toggle--hover--BorderWidth} - #{$pf-c-dropdown__toggle--BorderWidth}));
+      padding-right: calc(#{$pf-c-dropdown__toggle--PaddingRight} - (#{$pf-c-dropdown__toggle--hover--BorderWidth} - #{$pf-c-dropdown__toggle--BorderWidth}));
```
@srambach @mattnolting Are we using an inner shadow instead of a calculation on the padding to show active states on btns, dropdown and inputs?

andresgalante commented on src/site/workspace.scss:
```diff
@@ -208,4 +208,22 @@
+  }
+}
+
+.is-alignRight {
```
we don't use camel case for class names.

jgiardino commented:
I just rebased and updated the variables. I also made these updates based on feedback:

- rename handlebar properties and files based on new guidelines
- rename classes based on guidelines
- add missing `role="menuitem"` attribute
- apply expanded styling to toggle when active
- include `>` for modifier classes that affect child elements
- add z-index variables
- add resets for font size, font weight, and line height
- use inset box shadow to create the borders


I left the arrow as-is for now. I checked a few other design systems to see whether they include the icon in the html or css, and this is what I found:

- In HTML:
  - [bulma](https://bulma.io/documentation/components/dropdown/)
  - [polaris (form select)](https://polaris.shopify.com/components/forms/select#navigation)
  - [carbon](http://carbondesignsystem.com/components/dropdown/code)
  - [lightning](https://www.lightningdesignsystem.com/components/combobox/)
  - [bootstrap 3](https://getbootstrap.com/docs/3.3/components/#btn-dropdowns)
- In CSS:
  - [bootstrap 4](http://getbootstrap.com/docs/4.1/components/dropdowns/)
  - [foundation](https://foundation.zurb.com/sites/docs/dropdown-menu.html)

andresgalante commented:
Hi @jgiardino I think we should merge this and then decide about the arrow. I am for having it on the HTML.

@dgutride How do you want to handle situations like this one, should we merge it and open an issue?

We've changed the notation of modifier classes to me `-m` instead of `is`.  I'll make that change when I do that story.

dgutride commented:
@andresgalante  - I agree with that - having smaller issues after we get the main work done is really great for getting momentum built on these components
